actionview/lib/action_view/helpers/tag_helper.rb

```diff
@@ -121,6 +121,26 @@ def method_missing(called, *args, **options, &block)
           end
       end
 
+      class TokenList < Set # :nodoc:
+        def initialize(template, values)
+          @template = template
+          super(values.flat_map { |value| value.to_s.split(" ") })
+        end
+
+        def +(other)
+          TokenList.new(@template, super(Array.wrap(other)))
+        end
+
+        def to_s
+          @template.safe_join(to_a, " ")
+        end
+
+        def to_str
+          @template.safe_join([nil, to_a], " ")
```

Comment:
It wasn't immediately obvious to me why we needed `nil` here (I ran the tests to discover that it is so we will get the leading space when we do `"foo" + class_names("bar", "baz")`). What do you think about:

```suggestion
          " #{to_s}"
```

Reply:
👏 phew, much better.